### PR TITLE
[Cypress runner] Run Cypress

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
PENDING REVIEW

### What does this PR accomplish?
- Extracts the logic of how we run Cypress tests from the runner into a separate file
- Replaces our custom Cypress runner with the more suitable [Cypress Module API](https://docs.cypress.io/guides/guides/module-api)
- Implements the custom logic based on our custom `--open` flag that determines whether Cypress should start in [`cypress.run`](https://docs.cypress.io/guides/guides/module-api#cypress-run) or [`cypress.open`](https://docs.cypress.io/guides/guides/module-api#cypress-open) mode
- Adds a new way to conditionally run tests that rely on QA Docker databases by requiring `QA_DB_ENABLED` env set to true
    - This step would not even be needed if we manage to move `metabase-db` specs inside the main ones and run them together
    - That requires running [all 3 QA Docker images](https://github.com/metabase/metabase-qa) in CircleCI at the same time (the scope for a next PR)
    - Originally, I tried to [conditionally ignore test](https://github.com/metabase/metabase/pull/17099/commits/3f9b3bad3bc5f03f44e616d42481bd0d4e89937d#diff-2856eb2a1a98d3993ec3872ec449d66ddefe72846d18fb119dda596b252afe43R60) files in the runner itself, but [that option is not supported there](https://docs.cypress.io/guides/guides/module-api#Options), so I had to move it to the `frontend/test/__support__/e2e/cypress-plugins.js` file

### Why this change?
Our old Cypress runner was based on a completely custom implementation. It was `spawn`ing Node processes for both snapshots and actual tests. We had to return Promises and were missing on incredibly rich `results` object - both of which come for free with the Cypress Module API.
Please read more about the reasoning behind this change [here](https://github.com/metabase/metabase/issues/17005).

### How to test this?
``` bash
# should run only the files in `frontend/test/metabase/scenarios/admin/` folder
yarn test-cypress-no-build --folder admin

# should run only this one spec
yarn test-cypress-no-build --spec frontend/test/metabase/scenarios/native/native.cy.spec.js

# should run only the files in `frontend/test/metabase-db/mongo/` folder
QA_DB_ENABLED=true
yarn test-cypress-no-build --folder mongo

# should open Cypress GUI (open mode)
yarn test-cypress-open
```

Any of these commands should work.

You can also try **any** of the checks provided in https://github.com/metabase/metabase/pull/17010 (broken config, busted backend, failure in one of the tests, etc). Those steps and tests will be added explicitly in this description.

# Test scenarios

1. `yarn test-cypress-no-build --folder native`
- should run only the tests in the `frontend/test/metabase/scenarios/native/` folder
- should ignore tests from `frontend/test/metabase-db/` folder
<details>
  <summary>Click to reveal results</summary>

```
  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    6.8.0                                                                              │
  │ Browser:    Electron 87 (headless)                                                             │
  │ Specs:      7 found (frontend/test/metabase/scenarios/native/data_ref.cy.spec.js, frontend/tes │
  │             t/metabase/scenarios/native/native_subquery.cy.spec.js, frontend/test/metabase/sce │
  │             narios/native/native.cy.spec.js, frontend/test/metabase/scenarios/native/reproduct │
  │             ions/16886...)                                                                     │
  │ Searched:   frontend/test/metabase/scenarios/native/**/*.cy.spec.js                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

</details>

---

2. `yarn test-cypress-no-build --folder mongo`
- we forgot to set `QA_DB_enabled` env to true
- the whole process should fail

<details>
  <summary>Click to reveal results</summary>

```
Can't run because no spec files were found.

We searched for any files matching this glob pattern:

frontend/test/metabase-db/mongo/**/*.cy.spec.js

Relative to the project root folder:

/path/to/metabase
Could not find Cypress test run results
Cleaning up...
Stopped backend (host=http://localhost:4000 dbKey=/var/folders/2l/1x4y32qj12qfl7vs8b5d9ys80000gn/T/metabase-test-53771-0.db)
error Command failed with exit code 1.
```
</details>

---

3. `QA_DB_ENABLED=true yarn test-cypress-no-build --folder mongo`
- should run only the tests from the `./frontend/test/metabase-db/mongo/` folder

<details>
  <summary>Click to reveal results</summary>

```
  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    6.8.0                                                                              │
  │ Browser:    Electron 87 (headless)                                                             │
  │ Specs:      4 found (frontend/test/metabase-db/mongo/add.cy.spec.js, frontend/test/metabase-db │
  │             /mongo/line.cy.spec.js, frontend/test/metabase-db/mongo/native.cy.spec.js, fronten │
  │             d/test/metabase-db/mongo/query.cy.spec.js)                                         │
  │ Searched:   frontend/test/metabase-db/mongo/**/*.cy.spec.js                                    │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```
</details>

---

4. `yarn test-cypress-no-build`
- should run all tests from `./frontend/test/metabase/scenarios/` folder
- should ignore tests from `./frontend/test/metabase-db/` folder

<details>
  <summary>Click to reveal results</summary>

```
  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    6.8.0                                                                              │
  │ Browser:    Electron 87 (headless)                                                             │
  │ Specs:      149 found (frontend/test/metabase/scenarios/admin/audit/ad-hoc.cy.spec.js, fronten │
  │             d/test/metabase/scenarios/admin/audit/auditing.cy.spec.js, frontend/test/metabase/ │
  │             scenarios/admin/audit/questions-audit.cy.spec.js, frontend/test/metabase/scenarios │
  │             /admin/datab...)                                                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```
</details>

---

5. `yarn test-cypress-open`
- should open Cypress GUI
- searching for "mongo", "postgres" or "mysql" should yield "No results", because we told Cypress to ignore those folders

<details>
  <summary>Click to reveal results</summary>

image

</details>

---

6. Missing plugins file
- All the steps prior to "Starting Cypress" should pass
- "Starting Cypress" process should exit with the code 1 and clean up the backend

<details>
  <summary>Click to reveal results</summary>

```
Starting Cypress
The plugins file is missing or invalid.

Your `pluginsFile` is set to `/path/to/metabase/frontend/test/__support__/e2e/cypess-plugins.js`, but either the file is missing, it contains a syntax error, or threw an error when required. The `pluginsFile` must be a `.js`, `.ts`, or `.coffee` file.

Or you might have renamed the extension of your `pluginsFile`. If that's the case, restart the test runner.

Please fix this, or set `pluginsFile` to `false` if a plugins file is not necessary for your project.
Could not find Cypress test run results
Cleaning up...
Stopped backend (host=http://localhost:4000 dbKey=/var/folders/2l/1x4y32qj12qfl7vs8b5d9ys80000gn/T/metabase-test-56444-0.db)
error Command failed with exit code 1.
```

</details>

---

7. Broken or missing config file
- All the steps prior to "Starting Cypress" should pass
- "Starting Cypress" process should exit with the code 1 and clean up the backend

<details>
  <summary>Click to reveal results</summary>

```
Starting Cypress
Could not find a Cypress configuration file, exiting.

We looked but did not find a frontend/test/__support__/e2e/cypess.json file in this folder: /path/to/metabase
Could not find Cypress test run results
Cleaning up...
Stopped backend (host=http://localhost:4000 dbKey=/var/folders/2l/1x4y32qj12qfl7vs8b5d9ys80000gn/T/metabase-test-60015-0.db)
error Command failed with exit code 1.
```

### CODE REVIEW

File: frontend/test/__support__/e2e/cypress-plugins.js [hunk 1]
```diff
@@ -54,6 +56,9 @@ module.exports = (on, config) => {
    **                          CONFIG                                **
    ********************************************************************/
 
+  if (!isQaDatabase) {
+    config.ignoreTestFiles = "**/metabase-db/**";
+  }
   config.env.HAS_ENTERPRISE_TOKEN = hasEnterpriseToken;
```
Comment (nemanjaglumac) [inline frontend/test/__support__/e2e/cypress-plugins.js:63]: If you're interested in observing the resolved config, you can `console.log(config)` here.

Comment (nemanjaglumac) [reply frontend/test/__support__/e2e/cypress-plugins.js:63]: For the `yarn test-cypress-open`, you can access the resolved config by clicking on the "Settings" tab and then opening "Configuration"
image

File: frontend/test/__runner__/cypress-runner-run-tests.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    {},
+    defaultConfig,
+    reporterConfig,
+    userArgs,
+  );
+
+  try {
+    const { status, message, totalFailed, failures } = isOpenMode
+      ? await cypress.open(finalConfig)
+      : await cypress.run(finalConfig);
```
Comment (nemanjaglumac) [inline frontend/test/__runner__/cypress-runner-run-tests.js:77]: Also, for anyone interested in how the `results` object look like, feel free to console log the result of `await cypress.run(finalConfig)` or take a look at the example [here](https://docs.cypress.io/guides/guides/module-api#Results).

**Comment by nemanjaglumac** [comment]
Do we even need the `launch` function anymore?
https://github.com/metabase/metabase/blob/2e06b6c40ee1c40ae90e1ff926f9c65d0e44f435/frontend/test/__runner__/run_cypress_tests.js#L32

```javascript
const launch = () =>
  init()
    .then(cleanup)
    .catch(e => {
      console.error(e);
      cleanup(1);
    });

launch();
```

**Review by gusaiani — APPROVED** [review]
Tested the examples you mention in the description ✅

Code looks good to me.

The README files you touched.
What if we added links to them from the main README?
Seems like the majority of people seeing the repo, including Metabase engineers, would overlook README files in nested directories.

**Comment by nemanjaglumac** [comment]
> The README files you touched.
> What if we added links to them from the main README?
> Seems like the majority of people seeing the repo, including Metabase engineers, would overlook README files in nested directories.

Tbh, I'd get rid of these readme files all together. They are there temporarily. With one of the next refactors, they'll be gone.

**Review by ariya — APPROVED** [review]
LGTM. Let's ship this :+1:

**Pull request merged**